Add docs for compatibility changes between 4.4 and 5.1 (#1989)
emqtt/emq-docs-cn: en_US/changes/breaking-changes-5.1.0.md

:::

## Summary

Compared to EMQX 4.4, the upgrade to EMQX 5.1 introduces significant changes, particularly in terms of various concepts and mechanisms, surpassing the level of changes seen in previous upgrades from EMQX 2.x to EMQX 3.x and from EMQX 3.x to EMQX 4.x.

In conclusion, there are several points that need to be noted:

1. There are significant changes in **Configurations and HTTP APIs**. Existing configurations and code that rely on these interfaces will require migration.
2. **Core MQTT protocol functionalities**, including Pub/Sub, Retainer, and Shared Subscription, remain fully compatible with client programs. However, there may be slight changes in the management interface.
3. Other functionalities related to Authentication, Authorization, Data Integration, and Protocol Access will require migration based on their respective functions.
4. Certain concepts have undergone changes. For example, new versions of **Plugins** have been introduced, which differ significantly from the old versions. The concept of **Modules** in the old version has been completely removed.

> 5. Pay attention to the **removal of certain features**, such as `mcast` for Cluster discovery and `EMQX Bridges` for Data integration, among others.

Reviewer: > and `EMQX Bridges` for Data integration

This is confusing. What does `EMQX Bridges` mean?

## HTTP APIs

- Port 8081 has been closed, and all API requests now use port 18083.
- It is no longer supported to use the username of the Dashboard, such as admin:public, to request the HTTP API. Instead, an API Key must be used.

## Configuration Files

- Format:

  - EMQX 4.x: Flat format with `path.to.key = value`.
  - EMQX 5.1: Supports nested format with `path{to{ key = value }}`.
- Sources:
  - EMQX 4.x:
    - Multiple files, such as `emqx.conf`, `listeners.conf`, `zones.conf`, etc.
    - Dynamic updates are stored in Mnesia. Once enabled, files cannot be updated for config changes.
  - EMQX 5.1:
    - `emqx.conf` for static configs.
    - `cluster.hocon` for dynamic updates.

## Log File Format

> Log files follow the same format, but the log content style has slight differences. The log format in EMQX 5.1 is more indexer-friendly.

Reviewer: > Log files follow the same format

vs 

> The log format in EMQX 5.1 is more indexer-friendly

I think saying "Log files follow the same format" is not exactly correct in this context.
In v5 we have the option to use either flat log file format as in 4.4, or structured json format, which is more indexer-friendly.

- MongoDB

  - The storage schema has changed.

  - In EMQX 4.4, resulting documents should contain topics lists by action key, like in Redis or JWT: 

    ```
    {
      "publish": ["t1", "t2"],
      "subscribe": ["t3", "t4"],
      "pubsub": ["t5", "t6"]
    }
    ```

  - In EMQX 5.1, the documents should contain individual rules with `permission, action, topics` fields. Note that `topics` should be an array of topics.

## Data Integration

In EMQX 5.1, there have been conceptual improvements to data integration:

- Full compatibility for Rule and SQL templates is ensured.
- Most of the configuration item names and formats for resources and bridging have changed.
- The previous **Rule** -> **Action** -> **Resources** process has been modified to **Rule** -> **Bridges**.
- The functionality of **Modules/Message Publish** has been moved into the Bridges.
- The [**Save Offline Message**](https://docs.emqx.com/en/enterprise/v4.4/rule/offline_msg_to_redis.html)**,** [**Get Subscriptions**](https://docs.emqx.com/en/enterprise/v4.4/rule/get_subs_from_redis.html), and EMQX Bridge features have been removed.

> - Tablestore, DolphinDB, Lindorm, and SAP Event Mesh are not supported.

Reviewer: yet

For a complete compatibility report, see [Data Integration Incompatibility Between EMQX 5.1 with EMQX 4.4](./data-integration-4.4-to-5.1-incompatibility.md).

## Data Persistence

> - [MQTT Message Persistence](https://docs.emqx.com/en/enterprise/v4.4/backend/backend.html#mqtt-message-persistence) is not implemented in EMQX 5.0 a5.1.

Reviewer: ```suggestion
- [MQTT Message Persistence](https://docs.emqx.com/en/enterprise/v4.4/backend/backend.html#mqtt-message-persistence) is not implemented in EMQX 5.0 and 5.1. It's planned for the later versions.
```

## Gateway

> In EMQX 4.x, various protocols can be configured through corresponding Plugins and Modules (only for the enterprise edition). However, in E5.0, we have introduced a new concept called Gateway.

Reviewer: ```suggestion
In EMQX 4.x, various protocols can be configured through corresponding Plugins and Modules (only for the enterprise edition). However, in EMQX 5.0, we have introduced a new concept called Gateway.
```

> - It is **completely incompatible** from the configuration and management approaches perspective. E 5.0 has a brand new configuration format and management approach.

Reviewer: ```suggestion
- It is **completely incompatible** from the configuration and management approaches perspective. EMQX 5.0 has a brand new configuration format and management approach.
```
